### PR TITLE
[BEAM-9641] Support ZetaSQL DATE type as a Beam LogicalType

### PR DESCRIPTION
**Pull request by robinyqiu**
This PR adds support of all ZetaSQL ([BigQuery Standard SQL](https://cloud.google.com/bigquery/docs/reference/standard-sql/date_functions)) DATE functions to BeamSQL:

- CURRENT_DATE
- EXTRACT
- DATE (constructing DATE from DATETIME not supported)
- DATE_ADD
- DATE_SUB
- DATE_DIFF
- DATE_TRUNC
- FORMAT_DATE
- PARSE_DATE
- UNIX_DATE
- DATE_FROM_UNIX_DATE

* WEEK part not supported

r: @apilloud 
cc: @TheNeuralBit @kennknowles 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
What do you think about going ahead and defining the date logical type in `org.apache.beam.sdk.schemas.logicaltypes`? It would be useful in other contexts - for example it would give us something to map Avro's logical date type to (currently it is just overloaded with millis-instant onto DATETIME)

cc: @reuvenlax

**Comment by robinyqiu** [comment]
> What do you think about going ahead and defining the date logical type in `org.apache.beam.sdk.schemas.logicaltypes`? It would be useful in other contexts - for example it would give us something to map Avro's logical date type to (currently it is just overloaded with millis-instant onto DATETIME)

Done. Thanks for the suggestion. I made the `Date` type a public logical type in `org.apache.beam.sdk.schemas.logicaltypes` and added a layer of indirection by letting `SqlTypes.DATE` reference it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.time.LocalDate;
+import org.apache.beam.sdk.schemas.Schema;
+
+/**
+ * A date without a time-zone.
+ *
+ * <p>It cannot represent an instant on the time-line without additional information such as an
+ * offset or time-zone.
+ */
+public class Date implements Schema.LogicalType<LocalDate, Long> {
```
Comment (apilloud) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java:29]: If I'm reading the correctly, `LocalDate` is the in memory type (a struct) and `Long` is the wire format (an offset from epoch)? This conversion could be quite expensive. It appears the Calc nodes both take an offset in this case, when we start to think about performance we might need to change the in memory type to be offset based.

Comment (apilloud) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java:29]: (If changing the in memory type is going to be difficult in the future, consider doing that now.)

**Comment by apilloud** [comment]
Oops, forgot to include in my comments: ZetaSQL's range is much smaller than the underlying type, can you add a test or two for that?  How do out of range values fail? (Also worth asking, do we need any special treatment for boundary conditions (`LocalDate.MIN`, `LocalDate.MAX`)? Probably not for now.)

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 2]
```diff
@@ -427,17 +430,12 @@ private static Expression value(
...
-        } else if (logicalId.equals(DateType.IDENTIFIER)) {
-          value =
-              nullOr(
-                  value,
-                  Expressions.convert_(
-                      Expressions.divide(millisField, Expressions.constant(MILLIS_PER_DAY)),
-                      int.class));
+          return nullOr(
+              value, Expressions.convert_(Expressions.call(value, "getMillis"), int.class));
+        } else if (logicalId.equals(SqlTypes.DATE.getIdentifier())) {
```
[the remark this answers is not in the document]
Comment (TheNeuralBit) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java:437]: what about using a switch statement? Is there any style guidance on using switch on a String in java?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java:29]: That is unfortunate... but what in-memory type should we use instead? joda.time.LocalDate uses a millisecond long, do we want to add another joda dependency?

We could access the base type (wire format type) directly in SQL with [Row#getBaseValue](https://github.com/apache/beam/blob/7b890654e6bbfcab79a2f5677f1badd54bd444aa/sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java#L424), but unfortunately Rows store logical types as the input type (in memory format type), so that wouldn't actually avoid a conversion.

Comment (TheNeuralBit) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java:29]: I guess `java.sql.Date` is another option for a java type backed by millis.

Comment (apilloud) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/Date.java:29]: We should consider not using a JVM, it adds performance overhead too. 🤓 

I'm reasonably convinced the wire format is good and the conversion here is lossless, so if there isn't a easy drop-in replacement leave this as is.

**Comment by robinyqiu** [comment]
Ah, just realized that the previous comments were not sent out.

**Comment by robinyqiu** [comment]
Could you help trigger the tests again?

For the comment on range: Thanks for pointing it out. I overlooked this problem. I would like to create a separate PR to address it, along with range testing for other types as well.

**Comment by apilloud** [comment]
retest this please

**Comment by robinyqiu** [comment]
The failing test `SparkPortableExecutionTest.testExecution` should be unrelated to this change.

**Comment by apilloud** [comment]
Run Java PreCommit

**Comment by apilloud** [comment]
Run SQL Postcommit

**Comment by robinyqiu** [comment]
Rebased against master. Please run precommit tests again.

**Comment by apilloud** [comment]
retest this please

**Comment by apilloud** [comment]
retest this please

**Comment by robinyqiu** [comment]
Java PreCommit failed due to a build failure. Please help run again.

**Comment by TheNeuralBit** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
Something just occurred to me - are there any tests that use the DATE Type in an aggregation (e.g. MAX)?

I'd think that would run into the same issue I have in #11456

**Comment by apilloud** [comment]
Interesting question. You should probably add a test for JOIN as well, which will have a similar class of problems.

**Comment by robinyqiu** [comment]
> are there any tests that use the DATE Type in an aggregation (e.g. MAX)?

No. Thanks for bringing this up. I think it is likely to run into the problem.
